### DISCUSSION TITLE
How to do non-persisted computed values?

### DISCUSSION THREAD

**Discussion by NickEastNL** [mikro-orm/mikro-orm#5411] [Q&A]
I need to compute certain values based on stored properties before sending the data to the client. Usually I would use getters, which I assumed is what virtual properties were for. But they don't "exist" after fetching an item from the database. The only way I see them is when I use `wrap(entity).toJSON()`. But what if I need to find more than one entity? It doesn't seem right to first fetch the entities, then iterate through them to wrap each one up before sending them out.

The "fullName" example given in the docs is very unclear. It's the kind of effect I want, but it only seems to work after creating an entity, rather than fetching them. Then there's the matter of have to make the getters "optional" as otherwise `create` will complain they're missing.

So is there a way to have entities carry computed values directly when fetching them? Or do I _have_ to wrap them one by one?

**Comment by B4nan** [comment]
You can use getters, just add `persist: false` to the options as described in the docs

https://mikro-orm.io/docs/defining-entities#virtual-properties

The `fullName2` property in this example is a getter, sounds like that's exactly what you want.

**Reply by NickEastNL** [reply]
That's indeed what I want, but it's not working as I'd expect. If I have the following getter:
```js
@Property({ persist: false })
get myGetter() {
  return true;
}
```

```js
const entity = await this.entityRepository.findOne(id);

console.log(entity.myGetter); // undefined
```

But if I use `wrap().toJSON()`:
```js
const entity = await this.entityRepository.findOne(id);

console.log(wrap(entity).toJSON().myGetter); // true;
```

So if I need to find multiple entities, then I'd have to do the following in order for the getter to work later down the line:
```js
const entities = await this.entityRepository.find(ids);

const wrappedEntities = entities.map((v) => wrap(v).toJSON());
```

This doesn't _look_ bad, but it's not what I would expect, so I don't know if this is the only practical way to do it. I would have preferred it if I could just use the getter directly on the fetched entities without wrapping first. I assume returning the entities directly in an endpoint that returns JSON would serialize it just like above, but I need to access the getter in a Handlebars template.

**Reply by B4nan** [reply]
I'll check later today, it shouldn't behave like that, unless you have a matching database column, as getters are overridden with values from database.

You could try using `hydrate: false` or `trackChanges: false` to see if it helps, but it shouldn't be needed, if it is, its something to fix.

**Reply by B4nan** [reply]
I am not able to reproduce what you are saying, so complete repro welcome. Here is my take that is passing:

```ts
import {
  Entity,
  LoadStrategy,
  MikroORM,
  PrimaryKey,
  Property,
} from '@mikro-orm/sqlite';

@Entity()
class User {

  @PrimaryKey()
  id!: number;

  @Property()
  firstName!: string;

  @Property()
  lastName!: string;

  @Property({ persist: false })
  get fullName(): string {
    return `${this.firstName} ${this.lastName}`;
  }

  constructor(firstName: string, lastName: string) {
    this.firstName = firstName;
    this.lastName = lastName;
  }

}

let orm: MikroORM;

beforeAll(async () => {
  orm = await MikroORM.init({
    dbName: ':memory:',
    entities: [User],
    debug: true,
  });
  await orm.schema.createSchema();
});

afterAll(async () => {
  await orm.close(true);
});

test('virtual getter properties', async () => {
  const u = new User('first', 'last');
  expect(u.fullName).toBe('first last');
  await orm.em.persist(u).flush();
  orm.em.clear();

  const u2 = await orm.em.findOneOrFail(User, u);
  expect(u2.fullName).toBe('first last');
});
```

**Reply by NickEastNL** [reply]
I would have to apologize for wasting your time, this was a PEBCAK issue. I was inspecting the values incorrectly and therefore incorrectly assumed they didn't exist. Handlebars cannot access getters as they are not regular properties, and I never stopped and considered this until just now. Only when I implemented the code to return a JSON response did I realize this.

So I was doing it right by wrapping and serializing the entity to JSON, so that Handlebars can see the properties. Again, apologies
